src/network.cpp

```diff
@@ -0,0 +1,1106 @@
+#include "session/network.hpp"
+
+#include <oxenc/hex.h>
+#include <sodium/core.h>
+#include <sodium/crypto_sign_ed25519.h>
+
+#include <nlohmann/json.hpp>
+#include <oxen/log.hpp>
+#include <oxen/log/format.hpp>
+#include <oxen/log/ring_buffer_sink.hpp>
+#include <oxen/quic.hpp>
+#include <oxen/quic/opt.hpp>
+#include <random>
+#include <string>
+#include <string_view>
+
+#include "session/export.h"
+#include "session/network.h"
+#include "session/network_service_node.h"
+#include "session/network_service_node.hpp"
+#include "session/onionreq/builder.h"
+#include "session/onionreq/builder.hpp"
+#include "session/onionreq/key_types.hpp"
+#include "session/onionreq/response_parser.hpp"
+#include "session/util.hpp"
+
+using namespace session;
+using namespace oxen::quic;
+using namespace session::onionreq;
+using namespace std::literals;
+using namespace oxen::log::literals;
+
+namespace session::network {
+
+namespace {
+
+    inline auto log_cat = oxen::log::Cat("network");
+
+    class Timeout : public std::exception {};
+
+    const std::chrono::seconds default_timeout = 5s;
+
+    // The number of times a path can fail before it's replaced.
+    const uint16_t path_failure_threshold = 3;
+
+    // The number of times a snode can fail before it's replaced.
+    const uint16_t snode_failure_threshold = 3;
+
+    constexpr auto node_not_found_prefix = "Next node not found: "sv;
+    constexpr auto ALPN = "oxenstorage"sv;
+    const ustring uALPN{reinterpret_cast<const unsigned char*>(ALPN.data()), ALPN.size()};
+
+    std::array<uint8_t, 4> split_ipv4(std::string_view ip) {
+        std::array<uint8_t, 4> quad;
+        auto nums = split(ip, ".");
+        if (nums.size() != 4)
+            throw "Invalid IPv4 address";
+        for (int i = 0; i < 4; i++) {
+            auto end = nums[i].data() + nums[i].size();
+            if (auto [p, ec] = std::from_chars(nums[i].data(), end, quad[i]);
+                ec != std::errc{} || p != end)
+                throw "Invalid malformed IPv4 address";
+        }
+
+        return quad;
+    }
+}  // namespace
+
+Network::Network(const session::onionreq::ed25519_seckey ed25519_seckey) {
+    creds = GNUTLSCreds::make_from_ed_seckey(std::string(ed25519_seckey.view()));
+    endpoint = net.endpoint(Address{"0.0.0.0", 0}, opt::outbound_alpns{{uALPN}});
```

Comment:
```suggestion
    endpoint = net.endpoint(Address{"0.0.0.0", 0}, opt::alpns{{uALPN}});
```
Not strictly needed, but marginally less verbose.

Reply:
Updated